nexus/src/db/datastore/region.rs

```diff
@@ -177,30 +192,127 @@ impl DataStore {
                         )
                     })
                     .collect();
+
+                // Insert regions into the DB
                 let regions = diesel::insert_into(region_dsl::region)
                     .values(regions)
                     .returning(Region::as_returning())
                     .get_results(conn)?;
 
+                // Validate that the datasets (really, the zpools) have the
+                // space for these regions
+                for dataset in source_datasets.iter() {
+                    let zpool_id = dataset.pool_id;
+
+                    // Add up size used by all regions in all datasets in the
+                    // zpool
+                    let zpool_total_occupied_size: Option<
+                        diesel::pg::data_types::PgNumeric,
+                    > = region_dsl::region
+                        .filter(
+                            region_dsl::dataset_id.eq_any(
+                                dataset_dsl::dataset
+                                    .filter(dataset_dsl::pool_id.eq(zpool_id))
+                                    .filter(
+                                        dataset_dsl::size_used.is_not_null(),
+                                    )
+                                    .filter(dataset_dsl::time_deleted.is_null())
+                                    .filter(
+                                        dataset_dsl::kind
+                                            .eq(DatasetKind::Crucible),
+                                    )
+                                    .select(dataset_dsl::id),
+                            ),
+                        )
+                        .select(diesel::dsl::sum(
+                            region_dsl::block_size
+                                * region_dsl::blocks_per_extent
+                                * region_dsl::extent_count,
+                        ))
```

Comment:
I can see that my prior comment about "do this check for all dataset types, not just crucible" makes this more complicated. Could we perform this calculation using https://github.com/oxidecomputer/omicron/blob/48285cef35e61bb3216995ce17b91f136edc4e00/common/src/sql/dbinit.sql#L172-L174  instead of individual region allocations within a dataset?

Reply:
Done in 181c5d08